### PR TITLE
[stable backport] [UPG-NAT] Controlled NAT-Function

### PR DESCRIPTION
**Pull request by sergeymatov**
In order to allow UPG perform CG-NAT function following changes has been introduced:
To UPG:

- Ability to allocate and manage NAT Pools
- UE IP Address pool information IE to communicate NAT Pools to SMC
- BBF PFCP IEs to manage NAT Pool lookup and applying NAT to given UE IP
- UPG can manage NAT bindings configured for VPP NAT plugin

To NAT plugin:

- NAT bindings allocation/deletion mechanics exposed to UPG
- Binding lookup via bihash
- Override ED-NAT slow-path address and port allocation in respect to NAT binding created for given user address

This chain includes PFCP formatters fixes.

### CODE REVIEW

**Comment by sergeymatov** [comment]
> can we please concentrate on one PR, either this one or #122. Both are largely identical and any comment will apply to both anyway.

yes sure, please continue to comment #122

**Comment by ivan4th** [comment]
LGTM

File: upf/upf_pfcp_api.c [hunk 1]
```diff
@@ -1620,7 +1623,7 @@ handle_create_far (upf_session_t * sx, pfcp_create_far_t * create_far,
 	if (ISSET_BIT (far->forwarding_parameters.grp.fields,
 		       FORWARDING_PARAMETERS_BBF_APPLY_ACTION))
 	  {
-	    if (far->forwarding_parameters.bbf_apply_action &=
+	    if (far->forwarding_parameters.bbf_apply_action &
```
Comment (RoadRunnr) [inline upf/upf_pfcp_api.c]: this change is in the "checkstyle fixes" commit, but is actually a functional change

File: upf/upf.c [hunk 2]
```diff
@@ -109,7 +106,9 @@ upf_init_nat_addresses (upf_nat_pool_t * np, ip4_address_t start_addr,
 {
   u32 i = 0;
 
-  vec_reset_length (np->addresses);
+  vec_validate (np->addresses,
+		clib_net_to_host_u32 (end_addr.as_u32) -
+		clib_net_to_host_u32 (start_addr.as_u32) + 1);
```
Comment (RoadRunnr) [inline upf/upf.c:111]: I did suggest this change, but that does mean that you should turn off mind and copy what I proposed without thinking.

Simply replacing the` vec_reset_length` with `vec_validate` is neither sufficient nor correct.

Also, the commit should be folded into the NAT implementation commit.

**Review by RoadRunnr — CHANGES_REQUESTED** [review]

**Review by ivan4th — APPROVED** [review]
LGTM

File: upf/upf_pfcp_api.c (hunk 1 above)
Comment (sergeymatov) [reply upf/upf_pfcp_api.c]: commits updated

File: upf/upf_pfcp_api.c [hunk 3]
```diff
@@ -1634,11 +1637,6 @@ handle_create_far (upf_session_t * sx, pfcp_create_far_t * create_far,
 	      }
 	  }
 
-	//SMATOV: TEST
-	//u8 *pool_name = 0;
-	//pool_name = format (pool_name, "testing");
-	//handle_nat_binding_creation (sx, pool_name, response);
-	//vec_free (pool_name);
```
[the remark this answers is not in the document]
Comment (sergeymatov) [reply upf/upf_pfcp_api.c]: commits updated

**Pull request merged**